File: src/python/peak_rp.py
import sys
import numpy as np
import pandas as pd
from tqdm import tqdm

input_file = sys.argv[1]
output_file = sys.argv[2]
refseq_file = sys.argv[3]

L = 100_000  # Size of genomic inverval on either side of the TSS
delta = 10_000  # Decay distance


def calculate_single_gene_rp(sample_values, t_k, L=100000, mu=1.0):
    R_jk = 0
    for row in sample_values.itertuples():
        # Create a NumPy array of nucleotide positions
        i_values = np.arange(getattr(row, "start"), getattr(row, "end"))

        # Compute d, w_i, and s_ij values
        d_values = np.abs(i_values - t_k) / L
        w_i_values = 2 * np.exp(-mu * d_values) / (1 + np.exp(-mu * d_values))

        s_ij_values = getattr(row, "s_ij")

        # Compute R_jk value
        R_jk += np.sum(w_i_values * s_ij_values)

    return R_jk


def calculate_peak_rp(binary_values, refseq_genes, L=100000, mu=1.0):
    peak_rp = dict()

    for gene in tqdm(refseq_genes.itertuples(), total=refseq_genes.shape[0]):
        gene_id = getattr(gene, "gene_id")
        gene_chr = getattr(gene, "chr")
        t_k = getattr(gene, "tss")

        interval_start = max(t_k - L, 0)
        interval_end = t_k + L
        interval_values = binary_values[
            (binary_values["chr"] == gene_chr)
            & (binary_values["end"] >= interval_start)
            & (binary_values["start"] <= interval_end)
        ]

        R_jk = calculate_single_gene_rp(interval_values, t_k, L, mu)

        peak_rp[gene_id] = R_jk

    return peak_rp


binary_values = pd.read_csv(
    input_file,
    sep="\t",
    names=["chr", "start", "end", "s_ij"],
)

refseq_genes = pd.read_csv(
    refseq_file,
    sep="\t",
    names=["gene_id", "chr", "strand", "tss"],
)

mu = -np.log(L / (3 * delta))

peak_rp = calculate_peak_rp(binary_values, refseq_genes, L=L, mu=mu)

with open(output_file, "w") as f:
    for key in peak_rp.keys():
        f.write("%s,%s\n" % (key, peak_rp[key]))

File: src/cpp/peak_rp.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <unordered_map>
#include <cmath>
#include <algorithm>
#include <sstream>

// Import necessary libraries for handling CSV files
#include "csv.h" // Use a CSV parsing library, e.g., https://github.com/ben-strasser/fast-cpp-csv-parser

const int L = 100000;
const int delta = 10000;
const double mu = -std::log(static_cast<double>(L) / (3 * delta));

double calculate_single_gene_rp(const std::vector<std::tuple<int, int, double>> &sample_values, int t_k) {
    double R_jk = 0;
    for (const auto &[start, end, s_ij] : sample_values) {
        int adjusted_start = std::max(start, t_k - L);
        int adjusted_end = std::min(end, t_k + L);
        for (int i = adjusted_start; i < adjusted_end; ++i) {
            double d = std::abs(i - t_k) / static_cast<double>(L);
            double exp_neg_mu_d = std::exp(-mu * d);
            double w_i = 2 * exp_neg_mu_d / (1 + exp_neg_mu_d);
            R_jk += w_i * s_ij;
        }
    }
    return R_jk;
}


int main(int argc, char *argv[]) {
    if (argc != 4) {
        std::cerr << "Usage: " << argv[0] << " <input_file> <output_file> <refseq_file>" << std::endl;
        return 1;
    }

    const char *input_file = argv[1];
    const char *output_file = argv[2];
    const char *refseq_file = argv[3];

    // Read signal values (e.g., binarized TF binding, H3K27ac signal)
    io::CSVReader<4, io::trim_chars<' ', '\t'>, io::no_quote_escape<'\t'>> signal_values_reader(input_file);

    std::string chr;
    int start, end;
    double s_ij;
    std::unordered_map<std::string, std::vector<std::tuple<int, int, double>>> signal_values;

    while (signal_values_reader.read_row(chr, start, end, s_ij)) {
        signal_values[chr].push_back({start, end, s_ij});
    }

    // Read refseq genes
    io::CSVReader<4, io::trim_chars<' ', '\t'>, io::no_quote_escape<'\t'>> refseq_genes_reader(refseq_file);

    std::string gene_id, strand;
    int tss;
    std::vector<std::tuple<std::string, std::string, int>> refseq_genes;

    while (refseq_genes_reader.read_row(gene_id, chr, strand, tss)) {
        refseq_genes.push_back({gene_id, chr, tss});
    }


    // Calculate peak RP
    std::unordered_map<std::string, double> peak_rp;

    for (const auto &[gene_id, gene_chr, t_k] : refseq_genes) {
        int interval_start = std::max(t_k - L, 0);
        int interval_end = t_k + L;

        std::vector<std::tuple<int, int, double>> interval_values;
        for (const auto &row : signal_values[gene_chr]) {
            int start, end;
            double s_ij;
            std::tie(start, end, s_ij) = row;
            if (end >= interval_start && start <= interval_end) {
                interval_values.push_back(row);
            }
        }


        double R_jk = calculate_single_gene_rp(interval_values, t_k);
        peak_rp[gene_id] = R_jk;
    }


    // Write the output to a file
    std::ofstream outfile(output_file);

    if (!outfile.is_open()) {
        std::cerr << "Error opening output file: " << output_file << std::endl;
        return 1;
    }

    for (const auto &[key, value] : peak_rp) {
        outfile << key << "," << value << std::endl;
    }

    outfile.close();
    return 0;
}

   
